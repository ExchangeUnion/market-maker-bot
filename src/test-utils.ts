import { Config } from './config';
import { Level, Loggers } from './logger';

const testConfig = (): Config => {
  return {
    LOG_LEVEL: Level.Trace,
    CEX: 'Binance',
    CEX_API_KEY: '123',
    CEX_API_SECRET: 'abc',
    DATA_DIR: '',
    OPENDEX_CERT_PATH: `${__dirname}/../mock-data/tls.cert`,
    OPENDEX_RPC_HOST: 'localhost',
    OPENDEX_RPC_PORT: '1234',
    MARGIN: '0.06',
    BASEASSET: 'ETH',
    QUOTEASSET: 'BTC',
    CEX_BASEASSET: 'ETH',
    CEX_QUOTEASSET: 'BTC',
    TEST_CENTRALIZED_EXCHANGE_BASEASSET_BALANCE: '321',
    TEST_CENTRALIZED_EXCHANGE_QUOTEASSET_BALANCE: '123',
    TEST_MODE: true,
  };
};

const getLoggers = (): Loggers => {
  const mockLogger = {
    warn: () => {},
    info: () => {},
    verbose: () => {},
    debug: () => {},
    trace: () => {},
    error: () => {},
  };
  return ({
    global: mockLogger,
    centralized: mockLogger,
    opendex: mockLogger,
  } as unknown) as Loggers;
};

type TestError = {
  code: string | number;
  message: string;
};

export { getLoggers, testConfig, TestError };
