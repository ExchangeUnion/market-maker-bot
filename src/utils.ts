import moment from 'moment';
import { Observable, Subject } from 'rxjs';
import { take, tap } from 'rxjs/operators';
import { curry } from 'ramda';

/** Get the current date in the given dateFormat, if not provided formats with `YYYY-MM-DD hh:mm:ss.sss`.
 */
export const getTsString = (dateFormat?: string): string =>
  moment().format(dateFormat || 'YYYY-MM-DD hh:mm:ss.sss');

const SATOSHIS_PER_COIN = 10 ** 8;

/** Returns a number of satoshis as a string representation of coins with up to 8 decimal places. */
export const satsToCoinsStr = (satsQuantity: number): string => {
  return (satsQuantity / SATOSHIS_PER_COIN).toFixed(8).replace(/\.?0+$/, '');
};

/** Returns a number of coins as an integer number of satoshis. */
export const coinsToSats = (coinsQuantity: number): number => {
  return Math.round(coinsQuantity * SATOSHIS_PER_COIN);
};

export const getStartShutdown$ = (): Observable<unknown> => {
  const shutdown$ = new Subject();
  process.on('SIGINT', () => shutdown$.next());
  process.on('SIGTERM', () => shutdown$.next());
  return shutdown$.asObservable().pipe(take(1));
};

const debugObservable = (prefix: string, source: Observable<any>) => {
  return source.pipe(
    tap({
      next: v => console.log(`${prefix} next: ${v}`),
      error: e => console.log(`${prefix} error: ${e}`),
      complete: () => console.log(`${prefix} complete`),
    })
  );
};

const debugLog = curry(debugObservable);

export { debugLog };
