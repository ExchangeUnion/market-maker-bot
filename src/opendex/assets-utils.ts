import { BigNumber } from 'bignumber.js';
import { Logger } from '../logger';
import { GetBalanceResponse, TradingLimitsResponse } from '../proto/xudrpc_pb';
import { OpenDEXassetAllocation } from '../trade/info';
import { satsToCoinsStr } from '../utils';
import { errors } from './errors';

type LogAssetBalanceParams = {
  logger: Logger;
  assets: OpenDEXassetAllocation;
};

const logAssetBalance = ({ logger, assets }: LogAssetBalanceParams): void => {
  const {
    baseAssetBalance,
    baseAssetMaxInbound,
    baseAssetMaxOutbound,
    quoteAssetBalance,
    quoteAssetMaxInbound,
    quoteAssetMaxOutbound,
  } = assets;
  logger.trace(
    `Base asset balance ${baseAssetBalance.toFixed()} (maxinbound: ${baseAssetMaxInbound.toFixed()}, maxoutbound: ${baseAssetMaxOutbound.toFixed()}) and quote asset balance ${quoteAssetBalance.toFixed()} (maxinbound: ${quoteAssetMaxInbound.toFixed()}, maxoutbound: ${quoteAssetMaxOutbound.toFixed()}).`
  );
};

type ParseOpenDEXassetsParams = {
  balanceResponse: GetBalanceResponse;
  tradingLimitsResponse: TradingLimitsResponse;
  baseAsset: string;
  quoteAsset: string;
};

const parseOpenDEXassets = ({
  balanceResponse,
  tradingLimitsResponse,
  baseAsset,
  quoteAsset,
}: ParseOpenDEXassetsParams): OpenDEXassetAllocation => {
  const balancesMap = balanceResponse.getBalancesMap();
  const baseAssetBalances = balancesMap.get(baseAsset);
  if (!baseAssetBalances) {
    throw errors.BALANCE_MISSING(baseAsset);
  }
  const baseAssetBalance = new BigNumber(
    satsToCoinsStr(baseAssetBalances.getChannelBalance())
  );
  const quoteAssetBalances = balancesMap.get(quoteAsset);
  if (!quoteAssetBalances) {
    throw errors.BALANCE_MISSING(quoteAsset);
  }
  const quoteAssetBalance = new BigNumber(
    satsToCoinsStr(quoteAssetBalances.getChannelBalance())
  );
  const tradingLimitsMap = tradingLimitsResponse.getLimitsMap();
  const baseAssetLimits = tradingLimitsMap.get(baseAsset);
  if (!baseAssetLimits) {
    throw errors.TRADING_LIMITS_MISSING(baseAsset);
  }
  const baseAssetMaxOutbound = new BigNumber(
    satsToCoinsStr(baseAssetLimits.getMaxSell())
  ).plus(new BigNumber(satsToCoinsStr(baseAssetLimits.getReservedOutbound())));
  const baseAssetMaxInbound = new BigNumber(
    satsToCoinsStr(baseAssetLimits.getMaxBuy())
  ).plus(new BigNumber(satsToCoinsStr(baseAssetLimits.getReservedInbound())));
  const quoteAssetLimits = tradingLimitsMap.get(quoteAsset);
  if (!quoteAssetLimits) {
    throw errors.TRADING_LIMITS_MISSING(quoteAsset);
  }
  const quoteAssetMaxOutbound = new BigNumber(
    satsToCoinsStr(quoteAssetLimits.getMaxSell())
  ).plus(new BigNumber(satsToCoinsStr(quoteAssetLimits.getReservedOutbound())));
  const quoteAssetMaxInbound = new BigNumber(
    satsToCoinsStr(quoteAssetLimits.getMaxBuy())
  ).plus(new BigNumber(satsToCoinsStr(quoteAssetLimits.getReservedInbound())));
  return {
    baseAssetBalance,
    quoteAssetBalance,
    baseAssetMaxOutbound,
    baseAssetMaxInbound,
    quoteAssetMaxInbound,
    quoteAssetMaxOutbound,
  };
};

export {
  parseOpenDEXassets,
  logAssetBalance,
  LogAssetBalanceParams,
  ParseOpenDEXassetsParams,
};
