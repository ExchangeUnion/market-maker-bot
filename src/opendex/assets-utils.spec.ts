import { BigNumber } from 'bignumber.js';
import { Logger } from '../logger';
import { GetBalanceResponse, TradingLimitsResponse } from '../proto/xudrpc_pb';
import { getLoggers, testConfig } from '../test-utils';
import { satsToCoinsStr } from '../utils';
import { logAssetBalance, parseOpenDEXassets } from './assets-utils';

describe('OpenDEX.assets-utils', () => {
  test('parseOpenDEXassets success', () => {
    const config = testConfig();
    const baseAssetBalance = {
      getChannelBalance: () => 10000000,
    };
    const quoteAssetBalance = {
      getChannelBalance: () => 20000000,
    };
    const balancesMap = new Map();
    const baseAsset = config.BASEASSET;
    const quoteAsset = config.QUOTEASSET;
    balancesMap.set(baseAsset, baseAssetBalance);
    balancesMap.set(quoteAsset, quoteAssetBalance);
    const balanceResponse = ({
      getBalancesMap: () => balancesMap,
    } as unknown) as GetBalanceResponse;
    const limitsMap = new Map();
    const baseAssetLimits = {
      getMaxSell: () => 2500000,
      getMaxBuy: () => 2500000,
      getReservedInbound: () => 2500000,
      getReservedOutbound: () => 2500000,
    };
    const quoteAssetLimits = {
      getMaxSell: () => 5000000,
      getMaxBuy: () => 5000000,
      getReservedInbound: () => 5000000,
      getReservedOutbound: () => 5000000,
    };
    limitsMap.set(config.BASEASSET, baseAssetLimits);
    limitsMap.set(config.QUOTEASSET, quoteAssetLimits);
    const tradingLimitsResponse = ({
      getLimitsMap: () => limitsMap,
    } as unknown) as TradingLimitsResponse;
    expect(
      parseOpenDEXassets({
        baseAsset,
        quoteAsset,
        balanceResponse,
        tradingLimitsResponse,
      })
    ).toEqual({
      baseAssetBalance: new BigNumber(
        satsToCoinsStr(baseAssetBalance.getChannelBalance())
      ),
      baseAssetMaxOutbound: new BigNumber(
        satsToCoinsStr(
          baseAssetLimits.getMaxSell() + baseAssetLimits.getReservedOutbound()
        )
      ),
      baseAssetMaxInbound: new BigNumber(
        satsToCoinsStr(
          baseAssetLimits.getMaxBuy() + baseAssetLimits.getReservedInbound()
        )
      ),
      quoteAssetBalance: new BigNumber(
        satsToCoinsStr(quoteAssetBalance.getChannelBalance())
      ),
      quoteAssetMaxOutbound: new BigNumber(
        satsToCoinsStr(
          quoteAssetLimits.getMaxSell() + quoteAssetLimits.getReservedOutbound()
        )
      ),
      quoteAssetMaxInbound: new BigNumber(
        satsToCoinsStr(
          quoteAssetLimits.getMaxBuy() + quoteAssetLimits.getReservedInbound()
        )
      ),
    });
  });

  test('error baseAssetBalance', () => {
    const { BASEASSET, QUOTEASSET } = testConfig();
    const balancesMap = new Map();
    const balanceResponse = ({
      getBalancesMap: () => balancesMap,
    } as unknown) as GetBalanceResponse;
    const limitsMap = new Map();
    const tradingLimitsResponse = ({
      getLimitsMap: () => limitsMap,
    } as unknown) as TradingLimitsResponse;
    expect(() => {
      parseOpenDEXassets({
        baseAsset: BASEASSET,
        quoteAsset: QUOTEASSET,
        balanceResponse,
        tradingLimitsResponse,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test('error quoteAssetBalance', () => {
    const { BASEASSET, QUOTEASSET } = testConfig();
    const balancesMap = new Map();
    const baseAssetBalance = {
      getChannelBalance: () => 10000000,
    };
    balancesMap.set(BASEASSET, baseAssetBalance);
    const balanceResponse = ({
      getBalancesMap: () => balancesMap,
    } as unknown) as GetBalanceResponse;
    const limitsMap = new Map();
    const tradingLimitsResponse = ({
      getLimitsMap: () => limitsMap,
    } as unknown) as TradingLimitsResponse;
    expect(() => {
      parseOpenDEXassets({
        baseAsset: BASEASSET,
        quoteAsset: QUOTEASSET,
        balanceResponse,
        tradingLimitsResponse,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test('error baseAssetTradingLimits', () => {
    const { BASEASSET, QUOTEASSET } = testConfig();
    const balancesMap = new Map();
    const baseAssetBalance = {
      getChannelBalance: () => 10000000,
    };
    const quoteAssetBalance = {
      getChannelBalance: () => 10000000,
    };
    balancesMap.set(BASEASSET, baseAssetBalance);
    balancesMap.set(QUOTEASSET, quoteAssetBalance);
    const balanceResponse = ({
      getBalancesMap: () => balancesMap,
    } as unknown) as GetBalanceResponse;
    const limitsMap = new Map();
    const tradingLimitsResponse = ({
      getLimitsMap: () => limitsMap,
    } as unknown) as TradingLimitsResponse;
    expect(() => {
      parseOpenDEXassets({
        baseAsset: BASEASSET,
        quoteAsset: QUOTEASSET,
        balanceResponse,
        tradingLimitsResponse,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test('error quoteAssetLimits', () => {
    const { BASEASSET, QUOTEASSET } = testConfig();
    const balancesMap = new Map();
    const baseAssetBalance = {
      getChannelBalance: () => 10000000,
    };
    const quoteAssetBalance = {
      getChannelBalance: () => 10000000,
    };
    balancesMap.set(BASEASSET, baseAssetBalance);
    balancesMap.set(QUOTEASSET, quoteAssetBalance);
    const balanceResponse = ({
      getBalancesMap: () => balancesMap,
    } as unknown) as GetBalanceResponse;
    const limitsMap = new Map();
    const baseAssetLimits = {
      getMaxSell: () => 2500000,
      getMaxBuy: () => 2500000,
      getReservedOutbound: () => 2500000,
      getReservedInbound: () => 2500000,
    };
    limitsMap.set(BASEASSET, baseAssetLimits);
    const tradingLimitsResponse = ({
      getLimitsMap: () => limitsMap,
    } as unknown) as TradingLimitsResponse;
    expect(() => {
      parseOpenDEXassets({
        baseAsset: BASEASSET,
        quoteAsset: QUOTEASSET,
        balanceResponse,
        tradingLimitsResponse,
      });
    }).toThrowErrorMatchingSnapshot();
  });

  test('logAssetBalance', () => {
    const logger = (getLoggers().global as unknown) as Logger;
    logger.trace = jest.fn();
    const baseAssetBalance = '1.23456789';
    const baseAssetMaxOutbound = '0.617283945';
    const baseAssetMaxInbound = '0.33';
    const quoteAssetBalance = '9.87654321';
    const quoteAssetMaxInbound = '0.55';
    const quoteAssetMaxOutbound = '0.44';
    const assets = {
      baseAssetBalance: new BigNumber(baseAssetBalance),
      baseAssetMaxOutbound: new BigNumber(baseAssetMaxOutbound),
      baseAssetMaxInbound: new BigNumber(baseAssetMaxInbound),
      quoteAssetBalance: new BigNumber(quoteAssetBalance),
      quoteAssetMaxInbound: new BigNumber(quoteAssetMaxInbound),
      quoteAssetMaxOutbound: new BigNumber(quoteAssetMaxOutbound),
    };
    logAssetBalance({
      logger,
      assets,
    });
    expect(logger.trace).toHaveBeenCalledTimes(1);
    expect(logger.trace).toHaveBeenCalledWith(
      expect.stringContaining(baseAssetBalance)
    );
    expect(logger.trace).toHaveBeenCalledWith(
      expect.stringContaining(quoteAssetBalance)
    );
  });
});
