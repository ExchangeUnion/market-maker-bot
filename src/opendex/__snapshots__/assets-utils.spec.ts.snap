// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenDEX.assets-utils error baseAssetBalance 1`] = `"Could not retrieve ETH balance"`;

exports[`OpenDEX.assets-utils error baseAssetTradingLimits 1`] = `"Could not retrieve ETH trading limits"`;

exports[`OpenDEX.assets-utils error quoteAssetBalance 1`] = `"Could not retrieve BTC balance"`;

exports[`OpenDEX.assets-utils error quoteAssetLimits 1`] = `"Could not retrieve BTC trading limits"`;
